drivers/console/semihost_console.c

```diff
@@ -0,0 +1,39 @@
+/*
+ *	Copyright (c) 2019 LuoZhongYao
+ *	SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <sys/printk.h>
+
+extern void __printk_hook_install(int (*fn)(int));
+extern void __stdout_hook_install(int (*fn)(int));
+
+static int semihost_console_out(int ch)
+{
+	static unsigned char c;
+
+	c = ch;
+	__asm__ __volatile__ (
+		"movs	r1, %0\n"
+		"movs	r0, #3\n"
+		"bkpt	0xab\n"
```

Comment:
Note that this trap mechanism to specific to Cortex-M and will not work on Cortex-A and -R devices.

For Cortex-A and -R, the trap encoding is active instruction set-dependent (i.e. T32, A32 and A64 all have different encoding).

![image](https://user-images.githubusercontent.com/11345418/69901276-bcc79100-13c2-11ea-8425-d132d8ef3a59.png)
(from "Semihosting for AArch32 and AArch64 Release 2.0")